File: backend/app/api/v1/routes/bookings.py
"""
Endpoints del proceso de compra y emisión de boletos (RF-07, RF-02).

El endpoint `POST /v1/bookings/process` consolida en una sola
transacción atómica:

- Validación de los bloqueos temporales del `token_sesion`.
- Upsert de pasajeros por `numero_documento`.
- Emisión de boletos con `codigo_qr` único, vinculados al
  `id_usuario` autenticado (o NULL si el comprador es guest).
- Registro del pago en estado 'completado'.
- Conversión de los bloqueos ('convertido').

**RF-02 (guest checkout):** la autenticación es opcional. Si el
cliente envía Bearer token, el boleto queda asociado a su cuenta.
Si no envía token, el boleto se emite como guest (id_usuario NULL)
usando el `email_contacto` del payload como identificador.
"""

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import decode_token
from app.schemas.booking_schema import (
    BookingProcessRequest,
    BookingProcessResponse,
)
from app.services.booking_service import BookingService

router = APIRouter()


def _resolve_optional_user_id(request: Request) -> Optional[int]:
    """
    Devuelve el `id_usuario` si el request trae un Bearer token válido;
    None en caso contrario (flujo guest, RF-02).
    """
    auth = request.headers.get("authorization") or request.headers.get("Authorization")
    if not auth or not auth.lower().startswith("bearer "):
        return None
    token = auth.split(" ", 1)[1].strip()
    if not token:
        return None
    try:
        payload = decode_token(token)
    except Exception:
        return None
    sub = payload.get("sub")
    try:
        return int(sub)
    except (TypeError, ValueError):
        return None


@router.post(
    "/process",
    response_model=BookingProcessResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Procesar compra de boletos (RF-07, RF-02 guest)",
    tags=["Bookings"],
)
async def process_booking(
    payload: BookingProcessRequest,
    request: Request,
    db: Session = Depends(get_db),
) -> BookingProcessResponse:
    """
    Procesa el checkout completo de los asientos reservados.

    **Autenticación opcional:** si el cliente envía `Authorization:
    Bearer <jwt>`, el boleto se vincula a su `id_usuario`. Si NO
    envía token, el proceso se ejecuta como guest (RF-02): el
    `email_contacto` del payload se usa como dato de contacto y
    `id_usuario` queda NULL en la tabla `boletos`.

    Devuelve `201 CREATED` con el código de reserva, los boletos
    emitidos (cada uno con su `codigo_qr`), el resumen del pago y
    el total. Si algún asiento no tiene un bloqueo activo para el
    `token_sesion`, ya está ocupado, o no pertenece al bus del
    viaje, se devuelve `409 CONFLICT` y la transacción se revierte.
    """
    current_user_id = _resolve_optional_user_id(request)
    service = BookingService(db)
    try:
        result = service.process_booking(
            payload, id_usuario=current_user_id
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(exc),
        )
    return BookingProcessResponse(**result)

File: frontend-client/src/api/bookings.js
import axiosInstance from './axiosInstance'

function unwrapError(err) {
  if (err && typeof err === 'object' && 'message' in err) {
    return err
  }
  return new Error('Error inesperado al comunicarse con el servidor.')
}

/**
 * POST /v1/bookings/process
 * Procesa el checkout completo: valida bloqueos del token_sesion,
 * upserta pasajeros, emite boletos con QR y registra el pago.
 *
 * Soporta checkout como invitado (RF-02): si el cliente no envía
 * token Bearer, el backend emite el boleto con `id_usuario = NULL`
 * y usa el `email_contacto` del payload como identificador de
 * contacto. El interceptor de axios sólo añade el Authorization
 * header si hay token en localStorage.
 */
export async function processBookingRequest(payload) {
  try {
    const { data } = await axiosInstance.post('/bookings/process', payload)
    return {
      codigoReserva: data.codigo_reserva,
      idViaje: data.id_viaje,
      total: data.total,
      estado: data.estado,
      pago: {
        metodo: data.pago?.metodo,
        referenciaTransaccion: data.pago?.referencia_transaccion,
        montoTotal: data.pago?.monto_total,
        estado: data.pago?.estado,
      },
      boletos: (data.boletos || []).map((b) => ({
        idBoleto: b.id_boleto,
        idAsiento: b.id_asiento,
        numeroAsiento: b.numero_asiento,
        codigoQr: b.codigo_qr,
        precioFinal: b.precio_final,
        pasajero: b.pasajero,
      })),
    }
  } catch (err) {
    throw unwrapError(err)
  }
}
